README.md
# Flutter Reaction Button

[![pub package](https://img.shields.io/pub/v/flutter_reaction_button.svg)](https://pub.dartlang.org/packages/flutter_reaction_button)

Flutter button reaction it is fully customizable widget such as Facebook reaction button.

## Preview

<img src="https://github.com/GeekAbdelouahed/flutter-reaction-button/raw/doc/images/Preview.png"/>

## Demo

[![Flutter reaction button](https://img.youtube.com/vi/0REgZVVvTXg/default.jpg)](https://youtube.com/shorts/0REgZVVvTXg)

## Usage

[Include 'flutter_reaction_button' from Dart Pub.](https://pub.dartlang.org/packages/flutter_reaction_button)

```yaml
# pubspec.yaml

dependencies:
  flutter:
    sdk: flutter
  flutter_reaction_button: <last-version>
```

Next, import 'flutter_reaction_button.dart' into your dart code.

```dart
import 'package:flutter_reaction_button/flutter_reaction_button.dart';
```

## Parameters

| parameter                  | description                                                                           | default                                                                                                                                                                               |
| -------------------------- | ------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| OnReactionChanged          | triggered when reaction button value change                                           ||
| reactions                  | reactions appear in reactions box when long pressed on ReactionnButtonToggle or click on ReactionButton ||
| initialReaction            | Default reaction button widget                                                        | first item in reactions list |
| boxPosition                | Vertical position of reactions box relative to the button                             | VerticalPosition.TOP |
| boxHorizontalPosition      | Horizontal position of reactions box relative to the button                           | HorizontalPosition.START |
| boxOffset                  | Offset to reposition the box from the computed place                                  | Offset.zero |
| boxColor                   | Reactions box color                                                                   | Colors.white |
| boxElevation               | Reactions box elevation                                                               | 5 |
| boxRadius                  | Reactions box radius                                                                  | 50 |
| boxPadding                 | Reactions box padding                                                                 | const EdgeInsets.all(0) |
| boxDuration                | Reactions box show/hide duration                                                      | 200 milliseconds |
| shouldChangeReaction       | Should change initial reaction after selected one or not                               | true |
| itemScale                  | Scale ratio when item hovered                                                         | 0.3 |
| itemScaleDuration          | Scale duration while dragging                                                         | const Duration(milliseconds: 100) |

## LICENSE

```legal
MIT License

Copyright (c) 2019 Abdelouahed Medjoudja

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md
## 2.0.1+1

* Minor bugs fix
* Code improvements, Thanks to [Felix Gabler](https://github.com/felixgabler)

## 2.0.1

* Add ReactionContainer
* Bugs fix and code improvements, Thanks to [Felix Gabler](https://github.com/felixgabler)

## 2.0.0+3

* Upgrade flutter version

## 2.0.0+2

* Fix Offscreen Issue. Thanks to [rlee1990](https://github.com/rlee1990)

## 2.0.0+1

* Dragging improvement

## 2.0.0

* Change `FlutterReactionButton` to `ReactionButton`
* Change `FlutterReactionButtonCheck` to `ReactionButtonToggle`
* Change `Function(Reaction, int, bool) onReactionChanged` to `Function(T?, bool) onReactionChanged`
* Change `Function(Reaction, int) onReactionChanged` to `Function(T?) onReactionChanged`
* Dynamic scale depending on the hover position
* Dynamic horizontal alignment
* Scrollable position
* Minor bugs fixed

## 1.0.8

* Added support null safety

## 1.0.7+3

* Deprecated Reaction id

## 1.0.7+2

* Improvement box flow

## 1.0.7

* Remove reaction id
* Clean code

## 1.0.6+2

* Minor bugs fixed

## 1.0.6+1

* Minor bugs fixed

## 1.0.6

* Reaction title

## 1.0.5

* Items spacing
* Box padding

## 1.0.4

* Reactions box alignment

## 1.0.3

* Change state externally

## 1.0.2

* Code improvements

## 1.0.1

* Define reaction by id.
* Alternative to popup menu.
* Enable/disable reaction click.

## 1.0.0

* Add ripple effect.

## 0.1.3

* Selected rection index.

## 0.1.2+1

* Add Example.

## 0.1.2

* From plugin to packages.

## 0.1.1

* Reactions List @required.

## 0.1.0

* Update Description pubspec.yaml.

## 0.0.1

* Describe initial release.


lib/src/ui/reaction_container.dart
import 'package:flutter/material.dart';

import '../models/reaction.dart';
import '../utils/reactions_position.dart';
import 'reactions_box.dart';

class ReactionContainer<T> extends StatelessWidget {
  /// This triggers when reaction button value changed.
  final ValueChanged<T?> onReactionChanged;

  /// Previous selected reaction widget
  final Reaction<T>? selectedReaction;

  final List<Reaction<T>?> reactions;

  /// Vertical position of the reactions box relative to the button [default = VerticalPosition.TOP]
  final VerticalPosition boxPosition;

  /// Horizontal position of the reactions box relative to the button [default = HorizontalPosition.START]
  final HorizontalPosition boxHorizontalPosition;

  /// Reactions box color [default = white]
  final Color boxColor;

  /// Reactions box elevation [default = 5]
  final double boxElevation;

  /// Reactions box radius [default = 50]
  final double boxRadius;

  /// Reactions box show/hide duration [default = 200 milliseconds]
  final Duration boxDuration;

  /// Reactions box padding [default = const EdgeInsets.all(0)]
  final EdgeInsets boxPadding;

  /// Scale ratio when item hovered [default = 0.3]
  final double itemScale;

  /// Scale duration while dragging [default = const Duration(milliseconds: 100)]
  final Duration? itemScaleDuration;

  final Widget child;

  ReactionContainer({
    Key? key,
    required this.onReactionChanged,
    required this.reactions,
    this.selectedReaction,
    this.boxPosition = VerticalPosition.TOP,
    this.boxHorizontalPosition = HorizontalPosition.START,
    this.boxColor = Colors.white,
    this.boxElevation = 5,
    this.boxRadius = 50,
    this.boxDuration = const Duration(milliseconds: 200),
    this.boxPadding = const EdgeInsets.all(0),
    this.itemScale = .3,
    this.itemScaleDuration,
    required this.child,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      behavior: HitTestBehavior.translucent,
      onLongPressStart: (details) =>
          _showReactionsBox(context, details.globalPosition),
      child: child,
    );
  }

  void _showReactionsBox(BuildContext context, Offset buttonOffset) async {
    final reactionButton = await Navigator.of(context).push(
      PageRouteBuilder(
        opaque: false,
        transitionDuration: const Duration(milliseconds: 200),
        pageBuilder: (_, __, ___) {
          return ReactionsBox(
            buttonOffset: buttonOffset,
            buttonSize: Size.zero,
            reactions: reactions,
            verticalPosition: boxPosition,
            horizontalPosition: boxHorizontalPosition,
            color: boxColor,
            elevation: boxElevation,
            radius: boxRadius,
            duration: boxDuration,
            boxPadding: boxPadding,
            itemScale: itemScale,
            itemScaleDuration: itemScaleDuration,
          );
        },
      ),
    );

    if (reactionButton != null) onReactionChanged.call(reactionButton.value);
  }
}


example/lib/ui/common/item_container.dart
import 'package:flutter/material.dart';
import 'package:flutter_reaction_button/flutter_reaction_button.dart';

class ItemContainer extends StatefulWidget {
  final String imgPath;
  final List<Reaction<String>> reactions;

  const ItemContainer(this.imgPath, this.reactions);

  @override
  _ItemContainerState createState() => _ItemContainerState();
}

class _ItemContainerState extends State<ItemContainer>
    with AutomaticKeepAliveClientMixin {
  String? _selectedReaction;

  @override
  bool get wantKeepAlive => true;

  @override
  Widget build(BuildContext context) {
    super.build(context);
    return Padding(
      padding: const EdgeInsets.symmetric(
        horizontal: 10,
        vertical: 5,
      ),
      child: AspectRatio(
        aspectRatio: 2,
        child: Stack(
          children: [
            Positioned.fill(
              child: Card(
                margin: EdgeInsets.zero,
                elevation: 2,
                clipBehavior: Clip.antiAlias,
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.all(
                    Radius.circular(5),
                  ),
                ),
                child: ReactionContainer<String>(
                  onReactionChanged: (String? value) {
                    setState(() {
                      _selectedReaction = value;
                    });
                    debugPrint('Selected value: $value');
                  },
                  reactions: widget.reactions,
                  child: Image.asset(
                    widget.imgPath,
                    height: 200,
                    width: double.infinity,
                    fit: BoxFit.cover,
                  ),
                ),
              ),
            ),
            PositionedDirectional(
              end: 5,
              bottom: 5,
              child: Container(
                height: 30,
                width: 30,
                decoration: BoxDecoration(
                  color: Colors.black.withOpacity(.35),
                  shape: BoxShape.circle,
                  boxShadow: [
                    BoxShadow(
                      color: Colors.black.withOpacity(.3),
                      offset: const Offset(0, 3),
                      blurRadius: 3,
                    )
                  ],
                ),
                child: _selectedReaction != null
                    ? widget.reactions
                        .firstWhere((value) => value.value == _selectedReaction)
                        .previewIcon
                    : const SizedBox(),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/ui/reaction_button.dart
import 'package:flutter/material.dart';

import '../models/reaction.dart';
import '../utils/extensions.dart';
import '../utils/reactions_position.dart';
import 'reactions_box.dart';

class ReactionButton<T> extends StatefulWidget {
  /// This triggers when reaction button value changed.
  final ValueChanged<T?> onReactionChanged;

  /// Default reaction button widget
  final Reaction<T>? initialReaction;

  final List<Reaction<T>> reactions;

  /// Offset to add to the placement of the box
  final Offset boxOffset;

  /// Vertical position of the reactions box relative to the button [default = VerticalPosition.TOP]
  final VerticalPosition boxPosition;

  /// Horizontal position of the reactions box relative to the button [default = HorizontalPosition.START]
  final HorizontalPosition boxHorizontalPosition;

  /// Reactions box color [default = white]
  final Color boxColor;

  /// Reactions box elevation [default = 5]
  final double boxElevation;

  /// Reactions box radius [default = 50]
  final double boxRadius;

  /// Reactions box show/hide duration [default = 200 milliseconds]
  final Duration boxDuration;

  /// Change initial reaction after selected one [default = true]
  final bool shouldChangeReaction;

  /// Reactions box padding [default = EdgeInsets.zero]
  final EdgeInsetsGeometry boxPadding;

  /// Spacing between the reaction icons in the box
  final double boxReactionSpacing;

  /// Scale ratio when item hovered [default = 0.3]
  final double itemScale;

  /// Scale duration while dragging [default = const Duration(milliseconds: 100)]
  final Duration? itemScaleDuration;

  ReactionButton({
    Key? key,
    required this.onReactionChanged,
    required this.reactions,
    this.initialReaction,
    this.boxOffset = Offset.zero,
    this.boxPosition = VerticalPosition.TOP,
    this.boxHorizontalPosition = HorizontalPosition.START,
    this.boxColor = Colors.white,
    this.boxElevation = 5,
    this.boxRadius = 50,
    this.boxDuration = const Duration(milliseconds: 200),
    this.shouldChangeReaction = true,
    this.boxPadding = EdgeInsets.zero,
    this.boxReactionSpacing = 0,
    this.itemScale = .3,
    this.itemScaleDuration,
  }) : super(key: key);

  @override
  _ReactionButtonState createState() => _ReactionButtonState<T>();
}

class _ReactionButtonState<T> extends State<ReactionButton<T>> {
  final GlobalKey _buttonKey = GlobalKey();

  Reaction? _selectedReaction;

  void _init() {
    _selectedReaction = widget.initialReaction;
  }

  @override
  void didUpdateWidget(ReactionButton<T> oldWidget) {
    super.didUpdateWidget(oldWidget);
    _init();
  }

  @override
  void initState() {
    super.initState();
    _init();
  }

  @override
  Widget build(BuildContext context) => GestureDetector(
        key: _buttonKey,
        behavior: HitTestBehavior.translucent,
        onTapDown: (details) => _showReactionsBox(details.globalPosition),
        onLongPressStart: (details) =>
            _showReactionsBox(details.globalPosition),
        child: (_selectedReaction ?? widget.reactions.first).icon,
      );

  void _showReactionsBox(Offset buttonOffset) async {
    final buttonSize = _buttonKey.widgetSize;
    final reactionButton = await Navigator.of(context).push(
      PageRouteBuilder(
        opaque: false,
        transitionDuration: const Duration(milliseconds: 200),
        pageBuilder: (_, __, ___) {
          return ReactionsBox(
            buttonOffset: buttonOffset,
            buttonSize: buttonSize,
            reactions: widget.reactions,
            verticalPosition: widget.boxPosition,
            horizontalPosition: widget.boxHorizontalPosition,
            color: widget.boxColor,
            elevation: widget.boxElevation,
            radius: widget.boxRadius,
            offset: widget.boxOffset,
            duration: widget.boxDuration,
            boxPadding: widget.boxPadding,
            reactionSpacing: widget.boxReactionSpacing,
            itemScale: widget.itemScale,
            itemScaleDuration: widget.itemScaleDuration,
          );
        },
      ),
    );

    if (reactionButton != null) _updateReaction(reactionButton);
  }

  void _updateReaction(Reaction<T> reaction) {
    widget.onReactionChanged.call(reaction.value);
    if (mounted && widget.shouldChangeReaction)
      setState(() {
        _selectedReaction = reaction;
      });
  }
}
